src/main/java/com/limechain/runtime/Runtime.java

```diff
@@ -1,34 +1,43 @@
 package com.limechain.runtime;
 
-import com.limechain.runtime.allocator.FreeingBumpHeapAllocator;
 import com.limechain.runtime.hostapi.WasmExports;
 import com.limechain.runtime.hostapi.dto.RuntimePointerSize;
 import com.limechain.runtime.version.RuntimeVersion;
-import lombok.Getter;
+import com.limechain.runtime.version.scale.RuntimeVersionReader;
+import com.limechain.utils.scale.ScaleUtils;
 import lombok.extern.java.Log;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.wasmer.Instance;
-import org.wasmer.Memory;
 import org.wasmer.Module;
 
-import java.nio.ByteBuffer;
 import java.util.logging.Level;
 
-import static com.limechain.runtime.RuntimeBuilder.getImports;
-
-@Getter
 @Log
 public class Runtime {
-    private RuntimeVersion version;
-    private final Instance instance;
-    private final int heapPages;
-    private final FreeingBumpHeapAllocator allocator;
-
-    public Runtime(Module module, int heapPages) {
-        this.heapPages = heapPages;
-        this.instance = module.instantiate(getImports(module, this));
-        this.allocator = new FreeingBumpHeapAllocator(getHeapBase());
+    // TODO: Figure out whether we'll actually need this field
+    Module module; // used to open/close all of its own instances
+    Context context;
+    Instance instance;
+
+    // TODO: Reconsider whether `Context` should be a constructor argument or (dependency injected)
+    //                       OR an argument of `Runtime::call()` (dependency parameterized)
+    //  The former necessitates the explicit passing of a context for every build of the runtime
+    //      (we can't build the underlying module only - we always instantiate and tie the instance to the context)
+    //    - In more lightweight cases (i.e. only build to fetch the version), we'll have to explicitly provide a "partial" context;
+    //      "partial" meaning with some of the services inside being null (or partially implemented)
+    //      since we know they are not going to be used
+    //  The latter necessitates the explicit re-instantiating of the underlying `org.wasmer.Instance`
+    //      for each RuntimeAPI call (they do that in Gossamer),
+    //      but allows for easy caching of the compiled wasm module (thus we can pass that around, e.g. in the BlockState)
+    //      and instantiate with a context only when invocation is necessary.
+    //      This might prove more fitting, for example,
+    //      when executing consecutive blocks (between which there's no runtime updates)
+    //      without rebuilding the module and by only changing the context...
+    Runtime(Module module, Context context, Instance instance) {
```

Comment:
That's a rather significant one. It's a core design decision. I'm not sure, however, what's most fit for our needs, so for now we go the easy way - we need both the `byte[]` source and `Context` for instantiation; so a `Runtime` instance is bound with it's context for its entire lifetime. This would mean that even if there were no runtime updates between blocks, we'd have to re-instantiate the entire module in case we need a change in the context.

It's possible that's not optimal at some point, so it's free-to-refactor territory.

Reply:
You can try with the former approach and for the start not even have different Contexts but rather just a single full one. We might not always need the full context but it would be a good test to see how performant it would be.

Reply:
In any case, I think this topic is fine as-is for now. I don't plan on refactoring in this regard any further.

However, I really don't want to lose the discussion on that topic, or the knowledge of the other alternatives. We okay with that tedious TODO or where do we delegate this long-term? (I want to delete the TODO as it's rather big for an inline comment)

Reply:
After brief discussion, the comment was extracted into TODO.md. Fixed in [ed6abd6](https://github.com/LimeChain/Fruzhin/commit/ed6abd6a10c64bf5fc488fe30b4e0b173191d08c), in the follow-up PR.